File: test_cp31_linearization.cpp
#include "CRM.hpp"
#include "CRM_DiffDynamics.hpp"
#include <iostream>
#include <iomanip>
#include <cstring>
#include <cmath>

using namespace CRMCatheterModel;

// Helper to compute L2 norm
double vector_norm(const double* v, int n) {
    double sum = 0.0;
    for (int i = 0; i < n; i++) {
        sum += v[i] * v[i];
    }
    return std::sqrt(sum);
}

// Extract Jacobian matrices A (6x6) and B (6x3) using VJP
bool extract_jacobians(const double* x_t, const double* u_t, double dt, double L_inserted,
                       const CRMForwardKinematicsData& FKParams,
                       double* A, double* B) {
    DynamicsStepResult result;

    // Forward pass
    std::memset(&result, 0, sizeof(DynamicsStepResult));
    int status = dynamics_forward(x_t, u_t, dt, L_inserted, FKParams, result);
    if (status != 0 || result.lu_rank < 6) {
        std::cerr << "ERROR: Forward pass failed (status=" << status
                  << ", rank=" << result.lu_rank << ")" << std::endl;
        return false;
    }

    // Extract Jacobians via VJP with canonical basis vectors
    for (int k = 0; k < 6; k++) {
        double grad_x_next[6] = {0.0};
        grad_x_next[k] = 1.0;

        double grad_x_t[6], grad_u_t[3];
        int lu_rank;
        double rel_residual;

        int bwd_status = dynamics_backward(result, grad_x_next, FKParams, grad_x_t, grad_u_t,
                                          &lu_rank, &rel_residual);

        if (bwd_status != 0 || lu_rank < 6) {
            std::cerr << "ERROR: Backward pass failed for e_" << k
                      << " (status=" << bwd_status << ", rank=" << lu_rank << ")" << std::endl;
            return false;
        }

        // VJP gives (∂x_next/∂x_t)^T e_k = row k of A
        // VJP gives (∂x_next/∂u_t)^T e_k = row k of B
        for (int i = 0; i < 6; i++) {
            A[k * 6 + i] = grad_x_t[i];  // Row k of A
        }
        for (int i = 0; i < 3; i++) {
            B[k * 3 + i] = grad_u_t[i];  // Row k of B
        }
    }

    return true;
}

// Test linearization accuracy at a single operating point
bool test_operating_point(const char* name, const double* x_t, const double* u_t,
                          double dt, double L_inserted,
                          const CRMForwardKinematicsData& FKParams) {
    std::cout << "Operating Point: " << name << std::endl;
    std::cout << "  x_t = [";
    for (int i = 0; i < 6; i++) {
        std::cout << x_t[i];
        if (i < 5) std::cout << ", ";
    }
    std::cout << "]" << std::endl;
    std::cout << "  u_t = [" << u_t[0] << ", " << u_t[1] << ", " << u_t[2] << "]" << std::endl;
    std::cout << std::endl;

    // Extract Jacobians
    double A[36], B[18];
    if (!extract_jacobians(x_t, u_t, dt, L_inserted, FKParams, A, B)) {
        std::cerr << "FAIL: Could not extract Jacobians" << std::endl;
        return false;
    }

    // Define perturbations (very small for highly nonlinear dynamics)
    double dx[6] = {5e-7, 5e-7, 5e-7, 5e-6, 5e-6, 5e-6};  // Very small perturbations
    double du[3] = {5e-5, 5e-5, 5e-5};

    double norm_dx = vector_norm(dx, 6);
    double norm_du = vector_norm(du, 3);

    std::cout << "  Perturbations:" << std::endl;
    std::cout << "    ||dx|| = " << norm_dx << std::endl;
    std::cout << "    ||du|| = " << norm_du << std::endl;
    std::cout << std::endl;

    // Compute f(x_t, u_t) (base point)
    DynamicsStepResult result_base;
    std::memset(&result_base, 0, sizeof(DynamicsStepResult));
    int status_base = dynamics_forward(x_t, u_t, dt, L_inserted, FKParams, result_base);
    if (status_base != 0 || result_base.lu_rank < 6) {
        std::cerr << "FAIL: Base forward pass failed" << std::endl;
        return false;
    }

    // Compute f(x_t + dx, u_t + du) (actual)
    double x_pert[6], u_pert[3];
    for (int i = 0; i < 6; i++) x_pert[i] = x_t[i] + dx[i];
    for (int i = 0; i < 3; i++) u_pert[i] = u_t[i] + du[i];

    DynamicsStepResult result_pert;
    std::memset(&result_pert, 0, sizeof(DynamicsStepResult));
    int status_pert = dynamics_forward(x_pert, u_pert, dt, L_inserted, FKParams, result_pert);
    if (status_pert != 0 || result_pert.lu_rank < 6) {
        std::cerr << "FAIL: Perturbed forward pass failed" << std::endl;
        return false;
    }

    // Linearized prediction: x_next ≈ f(x_t, u_t) + A*dx + B*du
    double x_linear[6];
    for (int i = 0; i < 6; i++) {
        double A_dx = 0.0;
        for (int j = 0; j < 6; j++) {
            A_dx += A[i * 6 + j] * dx[j];
        }
        double B_du = 0.0;
        for (int j = 0; j < 3; j++) {
            B_du += B[i * 3 + j] * du[j];
        }
        x_linear[i] = result_base.x_next[i] + A_dx + B_du;
    }

    // Compute error
    double error_vec[6];
    for (int i = 0; i < 6; i++) {
        error_vec[i] = result_pert.x_next[i] - x_linear[i];
    }
    double error_norm = vector_norm(error_vec, 6);

    // Relative error
    double rel_error = error_norm / norm_dx;

    std::cout << "  Linearization Test:" << std::endl;
    std::cout << "    ||x_actual - x_linear|| = " << error_norm << std::endl;
    std::cout << "    Relative error (||error|| / ||dx||) = " << rel_error << std::endl;
    std::cout << std::endl;

    // Acceptance threshold: relative error < 0.1
    const double threshold = 0.1;
    bool pass = rel_error < threshold;

    if (pass) {
        std::cout << "  PASS (rel_error < " << threshold << ")" << std::endl;
    } else {
        std::cout << "  FAIL (rel_error >= " << threshold << ")" << std::endl;
    }
    std::cout << std::endl;

    return pass;
}

int main() {
    std::cout << std::setprecision(10);
    std::cout << "CP3.1 Linearization Validation Test (C++)" << std::endl;
    std::cout << "==========================================" << std::endl << std::endl;

    // Load parameters
    const char* param_file = "./catheterdata/CatheterParameterSet_1_dyn.txt";
    const char* config_file = "./catheterdata/CatheterSpatialConfiguration_1.txt";

    CRMCatheterModelParams CathParams = Load_CRMCatheterModelParams(param_file);
    CatheterConfiguration CathConfig = Load_CatheterConfiguration(config_file);

    // Setup FK params
    CRMForwardKinematicsData FKParams;
    FKParams.CathParams = &CathParams;
    FKParams.CathConfig = &CathConfig;
    FKParams.ContactMode = ContactModeType::FREE_TIP;
    FKParams.TipForce[0] = 0.0;
    FKParams.TipForce[1] = 0.0;
    FKParams.TipForce[2] = 0.0;
    FKParams.deltau0_initialguess[0] = 0.0;
    FKParams.deltau0_initialguess[1] = 0.0;
    FKParams.deltau0_initialguess[2] = 0.0;
    FKParams.IntegrationStepSize = 0.5;
    FKParams.FinalValueOnly = true;

    double dt = 0.01;
    double L_inserted = 50.0;

    // Operating Point 1: Rest
    double x_op1[6] = {0.0, 0.0, 0.0, 0.0, 0.0, 0.0};
    double u_op1[3] = {0.0, 0.0, 0.0};

    // Operating Point 2: Actuated
    double x_op2[6] = {0.0, 0.0, 0.0, 0.0, 0.0, 0.0};
    double u_op2[3] = {0.1, 0.0, 0.0};

    // Operating Point 3: Moving
    double x_op3[6] = {0.01, 0.0, 0.0, 0.1, 0.0, 0.0};
    double u_op3[3] = {0.1, 0.0, 0.0};

    // Test all operating points
    bool pass1 = test_operating_point("OP1 - Rest", x_op1, u_op1, dt, L_inserted, FKParams);
    bool pass2 = test_operating_point("OP2 - Actuated", x_op2, u_op2, dt, L_inserted, FKParams);
    bool pass3 = test_operating_point("OP3 - Moving", x_op3, u_op3, dt, L_inserted, FKParams);

    // Overall verdict
    std::cout << "========================================" << std::endl;
    std::cout << "Overall Results:" << std::endl;
    std::cout << "  OP1 (Rest):     " << (pass1 ? "PASS" : "FAIL") << std::endl;
    std::cout << "  OP2 (Actuated): " << (pass2 ? "PASS" : "FAIL") << std::endl;
    std::cout << "  OP3 (Moving):   " << (pass3 ? "PASS" : "FAIL") << std::endl;
    std::cout << std::endl;

    bool all_pass = pass1 && pass2 && pass3;
    if (all_pass) {
        std::cout << "CP3.1 PASS: All operating points validated" << std::endl;
        return 0;
    } else {
        std::cout << "CP3.1 FAIL: Some operating points failed" << std::endl;
        return 1;
    }
}

File: python/archieve/test_cp31_linearization.py
"""
CP3.1: Linearization Validation (Python + PyTorch)

Validates that:
1. PyTorch jacobian() extracts correct Jacobians A, B
2. PyTorch Jacobians match C++ Jacobians within numerical precision
3. Linearization error is acceptable for iLQR use
"""

import sys
import numpy as np
import torch
import crm_diff_py
from crm_dynamics_torch import dynamics_step, load_default_catheter_params


def extract_jacobians_pytorch(x_t, u_t, dt, L_inserted, params_dict):
    """
    Extract Jacobians A = ∂x_next/∂x_t and B = ∂x_next/∂u_t using PyTorch.

    Returns:
        A: (6, 6) numpy array
        B: (6, 3) numpy array
    """
    x_t_torch = torch.tensor(x_t, dtype=torch.float64, requires_grad=True)
    u_t_torch = torch.tensor(u_t, dtype=torch.float64, requires_grad=True)

    # Extract A = ∂x_next/∂x_t
    A = torch.autograd.functional.jacobian(
        lambda x: dynamics_step(x, u_t_torch, dt, L_inserted, params_dict),
        x_t_torch
    ).numpy()

    # Extract B = ∂x_next/∂u_t
    B = torch.autograd.functional.jacobian(
        lambda u: dynamics_step(x_t_torch, u, dt, L_inserted, params_dict),
        u_t_torch
    ).numpy()

    return A, B


def extract_jacobians_cpp(x_t, u_t, dt, L_inserted, params_dict):
    """
    Extract Jacobians using C++ dynamics_backward with VJP.

    Returns:
        A: (6, 6) numpy array
        B: (6, 3) numpy array
    """
    # Forward pass
    result = crm_diff_py.dynamics_forward(x_t, u_t, dt, L_inserted, params_dict)

    if result['status'] != 0 or result['lu_rank'] < 6:
        raise RuntimeError(f"Forward pass failed: status={result['status']}, rank={result['lu_rank']}")

    A = np.zeros((6, 6))
    B = np.zeros((6, 3))

    # Extract via VJP with canonical basis vectors
    for k in range(6):
        grad_x_next = np.zeros(6)
        grad_x_next[k] = 1.0

        bwd_result = crm_diff_py.dynamics_backward(result, grad_x_next, params_dict)

        if bwd_result['status'] != 0 or bwd_result['lu_rank'] < 6:
            raise RuntimeError(f"Backward pass failed: status={bwd_result['status']}, rank={bwd_result['lu_rank']}")

        # VJP gives row k of Jacobian
        A[k, :] = bwd_result['grad_x_t']
        B[k, :] = bwd_result['grad_u_t']

    return A, B


def test_operating_point(name, x_t, u_t, dt, L_inserted, params_dict):
    """
    Test linearization accuracy at a single operating point.

    Returns:
        True if all tests pass, False otherwise
    """
    print(f"Operating Point: {name}")
    print(f"  x_t = {x_t}")
    print(f"  u_t = {u_t}")
    print()

    # Extract Jacobians via PyTorch
    A_py, B_py = extract_jacobians_pytorch(x_t, u_t, dt, L_inserted, params_dict)

    # Extract Jacobians via C++
    A_cpp, B_cpp = extract_jacobians_cpp(x_t, u_t, dt, L_inserted, params_dict)

    # Compare PyTorch vs C++
    A_diff = np.linalg.norm(A_py - A_cpp, 'fro')
    B_diff = np.linalg.norm(B_py - B_cpp, 'fro')
    A_norm = np.linalg.norm(A_cpp, 'fro')
    B_norm = np.linalg.norm(B_cpp, 'fro')

    A_rel_err = A_diff / max(A_norm, 1e-12)
    B_rel_err = B_diff / max(B_norm, 1e-12)

    print("  Jacobian Comparison (PyTorch vs C++):")
    print(f"    ||A_py - A_cpp||_F = {A_diff:.6e}")
    print(f"    ||A_cpp||_F        = {A_norm:.6e}")
    print(f"    Relative error A   = {A_rel_err:.6e}")
    print()
    print(f"    ||B_py - B_cpp||_F = {B_diff:.6e}")
    print(f"    ||B_cpp||_F        = {B_norm:.6e}")
    print(f"    Relative error B   = {B_rel_err:.6e}")
    print()

    # Acceptance: relative error < 1e-6
    threshold = 1e-6
    pass_A = A_rel_err < threshold
    pass_B = B_rel_err < threshold

    if pass_A and pass_B:
        print(f"  PASS: PyTorch Jacobians match C++ (rel_err < {threshold})")
    else:
        print(f"  FAIL: PyTorch Jacobians do NOT match C++ (rel_err >= {threshold})")
    print()

    # Linearization accuracy test (same as C++)
    dx = np.array([5e-7, 5e-7, 5e-7, 5e-6, 5e-6, 5e-6])
    du = np.array([5e-5, 5e-5, 5e-5])

    norm_dx = np.linalg.norm(dx)
    norm_du = np.linalg.norm(du)

    # Compute f(x_t, u_t)
    result_base = crm_diff_py.dynamics_forward(x_t, u_t, dt, L_inserted, params_dict)
    if result_base['status'] != 0:
        print("  FAIL: Base forward pass failed")
        return False
    x_base = result_base['x_next']

    # Compute f(x_t + dx, u_t + du)
    x_pert = x_t + dx
    u_pert = u_t + du
    result_pert = crm_diff_py.dynamics_forward(x_pert, u_pert, dt, L_inserted, params_dict)
    if result_pert['status'] != 0:
        print("  FAIL: Perturbed forward pass failed")
        return False
    x_actual = result_pert['x_next']

    # Linearized prediction
    x_linear = x_base + A_cpp @ dx + B_cpp @ du

    # Error
    error = x_actual - x_linear
    error_norm = np.linalg.norm(error)
    rel_error = error_norm / norm_dx

    print("  Linearization Test:")
    print(f"    ||dx|| = {norm_dx:.6e}")
    print(f"    ||du|| = {norm_du:.6e}")
    print(f"    ||x_actual - x_linear|| = {error_norm:.6e}")
    print(f"    Relative error (||error|| / ||dx||) = {rel_error:.6e}")
    print()

    # Acceptance: relative error < 0.1
    lin_threshold = 0.1
    pass_lin = rel_error < lin_threshold

    if pass_lin:
        print(f"  PASS: Linearization accurate (rel_error < {lin_threshold})")
    else:
        print(f"  FAIL: Linearization inaccurate (rel_error >= {lin_threshold})")
    print()

    return pass_A and pass_B and pass_lin


def main():
    print("CP3.1 Linearization Validation Test (Python + PyTorch)")
    print("=======================================================")
    print()

    # Load parameters
    param_file = "./catheterdata/CatheterParameterSet_1_dyn.txt"
    config_file = "./catheterdata/CatheterSpatialConfiguration_1.txt"
    params_dict = load_default_catheter_params(param_file, config_file)

    dt = 0.01  # s
    L_inserted = 50.0  # mm (explicit, not from dataset)

    # Operating points
    x_op1 = np.array([0.0, 0.0, 0.0, 0.0, 0.0, 0.0])
    u_op1 = np.array([0.0, 0.0, 0.0])

    x_op2 = np.array([0.0, 0.0, 0.0, 0.0, 0.0, 0.0])
    u_op2 = np.array([0.1, 0.0, 0.0])

    x_op3 = np.array([0.01, 0.0, 0.0, 0.1, 0.0, 0.0])
    u_op3 = np.array([0.1, 0.0, 0.0])

    # Test all operating points
    pass1 = test_operating_point("OP1 - Rest", x_op1, u_op1, dt, L_inserted, params_dict)
    pass2 = test_operating_point("OP2 - Actuated", x_op2, u_op2, dt, L_inserted, params_dict)
    pass3 = test_operating_point("OP3 - Moving", x_op3, u_op3, dt, L_inserted, params_dict)

    # Overall verdict
    print("=" * 55)
    print("Overall Results:")
    print(f"  OP1 (Rest):     {'PASS' if pass1 else 'FAIL'}")
    print(f"  OP2 (Actuated): {'PASS' if pass2 else 'FAIL'}")
    print(f"  OP3 (Moving):   {'PASS' if pass3 else 'FAIL'}")
    print()

    all_pass = pass1 and pass2 and pass3
    if all_pass:
        print("CP3.1 PASS: All tests validated")
        return 0
    else:
        print("CP3.1 FAIL: Some tests failed")
        return 1


if __name__ == "__main__":
    sys.exit(main())
